compiler/src/dmd/file_manager.d

```diff
@@ -364,3 +464,212 @@ private bool readFromStdin(ref OutBuffer sink) nothrow
 
     assert(0);
 }
+
+/********************************************
+ * Read all the modules, and build one giant cache file out of them.
+ * Params:
+ *      pathPackage = the path/package to build a cache file for
+ * Returns:
+ *      true = failed
+ *      false = success
+ */
+bool writeSourceArchive(const(char)[] pathPackage)
+{
+    enum log = false;
+    if (log) printf("writeSourceArchive() %.*s\n", cast(int)pathPackage.length, pathPackage.ptr);
+
+    const(char)[] name = FileName.name(pathPackage);
+    const size_t nameStart = pathPackage.length - name.length;
+
+    Array!(const(char)*) fileNames;
+
+    void accumulate(scope const(char)[] filename) nothrow
+    {
+        if (log) printf("%.*s\n", cast(int)filename.length, filename.ptr);
+        fileNames.push(xarraydup(filename).ptr);
+    }
+
+    immutable string[4] exts = ["d", "di", "c", "i"];   // archive files with these extensions
+
+    const(char)[] dir_path = xarraydup(pathPackage);
+
+    if (find_files(dir_path.ptr, exts, &accumulate))
+        return true;
+
+    OutBuffer ar;  // ar will hold the contents of the archive file
+
+    SrcArchiveHeader srcArchiveHeader;
+    srcArchiveHeader.contentsOffset = cast(uint)SrcArchiveHeader.sizeof;
+    srcArchiveHeader.contentsLength = cast(uint)fileNames.length;
+
+    ar.write(&srcArchiveHeader, srcArchiveHeader.sizeof);
+
+    /* Allocate in ar the array of Content
+     */
+    size_t length = fileNames.length;
+    ar.allocate(length * Content.sizeof);
+
+    /* Temporary array of Content
+     */
+    Content[] contents = (cast(Content*)mem.xmalloc(length * Content.sizeof))[0 .. length];
+
+    /* Write the name strings to ar
+     */
+    foreach (i, ref content; contents[])
+    {
+        const(char)* fnp = fileNames[i];
+        const(char)[] fn = fnp[nameStart .. strlen(fnp)];       // slice off path/ prefix
+        if (log) printf("fn: %.*s\n", cast(int)fn.length, fn.ptr);
+
+        content.nameOffset = cast(uint)ar.length;
+        content.nameLength = cast(uint)fn.length;
+        //printf("1content.nameOffset[%lld]: %d nameLength: %d\n", i, content.nameOffset, content.nameLength);
+
+        ar.writeStringz(fn);
+    }
+
+    /* Read the files, and write their contents to ar, along with
+     * a terminating 4 bytes of 0 (to accommodate dchar source files)
+     */
+    OutBuffer fb;
+    foreach (i, ref content; contents[])
+    {
+        fb.reset();                          // recycle read buffer
+        File.read(fileNames[i][0 .. strlen(fileNames[i])], fb); // read file into fb
+        content.importOffset = cast(uint)ar.length;
+        content.importLength = cast(uint)fb.length; // don't include terminating 0
+        ar.write(fb.peekSlice());                   // append file contents to ar
+        ar.write32(0);                              // append 4 bytes of 0
```

Comment:
This is accounting for 2 trailing nulls for sentinel and 2 for lexer?

Reply:
It's accounting for a null termination if it is a UTF32 source file.